trieste/acquisition/function.py

```diff
@@ -201,6 +203,82 @@ def update_acquisition_function(
         return function
 
 
+class MonteCarloExpectedImprovement(SingleModelAcquisitionBuilder):
+    """
+    Builder for a Monte Carlo-based expected improvement function for use with a model where we have
+    to obtain a Monte Carlo estimate of the expected improvement. The "best" value is taken to be
+    the minimum of the posterior mean at observed points.
+    """
+
+    def __init__(self, sample_size: int):
+        """
+        :param sample_size: The number of samples for each batch of points.
+        :raise tf.errors.InvalidArgumentError: If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        super().__init__()
+
+        self._sample_size = sample_size
+
+    def __repr__(self) -> str:
+        """"""
+        return f"MonteCarloExpectedImprovement({self._sample_size!r})"
+
+    def prepare_acquisition_function(
+        self, dataset: Dataset, model: ProbabilisticModel
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer. Must be populated.
+        :param model: The model over the specified ``dataset``. Must have event shape [1] and must
+            be an instance of :class:`~trieste.models.gpflux.DeepGaussianProcess`.
+        :return: The estimated *expected improvement* acquisition function.
+        :raise ValueError (or InvalidArgumentError): If ``dataset`` is not populated, ``model``
+            does not have an event shape of [1] or is not an instance of
+            :class:`~trieste.models.gpflux.DeepGaussianProcess`.
+        """
+        mean_samples_op = getattr(model, "mc_posterior_mean", None)
```

Comment:
this should already throw an exception when the method doesn't exist, but you'll figure out details in tests (this one should be definitely tested)

Reply:
Do you think it's best to remove this then? It might be nice to have the custom error

Reply:
custom error is good but I'm thinking you need to create condition on this line rather than the one below

Reply:
I'm not sure I understand what you mean? Which line?